### PR TITLE
Adding artifactType filtering to the Referrers API

### PR DESCRIPTION
**Pull request by michaelb990**
Draft PR on top of #59 to add `artifactType` filtering in the Referrers API.

Pros:
* Enables registries and clients to more easily work with large numbers of references for an image.
* Prevents a use case which pushes large numbers of artifacts from impacting the performance of latency sensitive use cases.

Cons:
* Fragments registry behavior because not all registries may end up supporting this, so clients will still need a fall back.
* Only allows for filtering based on `artifactType`, not other fields or annotations.

### CODE REVIEW

**Review by nishakm — APPROVED** [review]

File: docs/proposals/PROPOSAL_E.md [hunk 1]
```diff
@@ -157,6 +163,51 @@ If a query results in no referrers found, an empty manifest list MUST be returne
 - The registry SHOULD return fewer than `n` results when the generated Index would exceed the recommended maximum manifest size.
 - The `Link` HTTP header is included in the response when additional results are available and is set to the URL for the next page of results.
 
+#### Filtering
+A limited ability to filter results based on `artifactType` will enable the referrers API to scale as the number of artifact use cases grows.
+ - The registry SHOULD allow for filtering based on `artifactType`.
+ - Clients MAY request results be filterd by adding the `artifactType` query parameter to requests to the referrers API.
```
Comment (imjasonh) [inline docs/proposals/PROPOSAL_E.md:169]:
```suggestion
 - Clients MAY request results be filtered by adding the `artifactType` query parameter to requests to the referrers API.
```

File: docs/proposals/PROPOSAL_E.md [hunk 2]
```diff
@@ -157,6 +163,51 @@ If a query results in no referrers found, an empty manifest list MUST be returne
...
+ - The value for `artifactType` MUST be the IANA media type of the artifact to be returned.
```
Comment (imjasonh) [inline docs/proposals/PROPOSAL_E.md:170]:
```suggestion
 - The value for `artifactType` MUST be the IANA media type of the artifact to be returned.
 - Clients SHOULD send only one `artifactType` query parameter, and registries SHOULD only interpret one when filtering, if a client sends multiple values.
```

Not totally sold on the wording, but I want to get out ahead of questions about filtering to include `A or B`

**Comment by jdolitsky** [comment]
Please rebase as https://github.com/opencontainers/wg-reference-types/pull/59 is in

File: docs/proposals/PROPOSAL_E.md [hunk 3]
```diff
@@ -179,6 +230,7 @@ For registries that do not support the `referrers` API, a tag MUST be pushed for
 - `<ref>`: the digest from the `refers` field (limit of 64 characters)
 - `<hash>`: the digest of this artifact (limit of 16 characters)
 - `<type>`: type of artifact for filtering (limit of 5 characters)
+  - The type field is based on a mapping from `artifactType` to a short string for the tag.
```
Comment (jdolitsky) [inline docs/proposals/PROPOSAL_E.md:233]: this is probably to go away depending on https://github.com/opencontainers/wg-reference-types/issues/72

File: docs/proposals/PROPOSAL_E.md [hunk 4]
```diff
@@ -157,6 +163,51 @@ If a query results in no referrers found, an empty manifest list MUST be returne
...
+      "digest": "sha256:a2a2a2...",
+      "annotations": [
+        "org.opencontainers.artifact.created": "2022-01-01T15:24:30Z",
+        "org.example.icecream.flavor": "vanilla"
+      ]
+    }
+  ],
+  "annotations": {
+    "artifactType": "application/vnd.example.icecream.v1"
+  }
```
Comment (jdolitsky) [inline docs/proposals/PROPOSAL_E.md:207]: should this be codified into the proposal?

Comment (sajayantony) [reply docs/proposals/PROPOSAL_E.md:207]: So from the discussion just an indication of the field without he values makes the response for unmatched artifacts cachable.

**Comment by sajayantony** [comment]
As a part of the response are all annotations expected to be returned in the `referrers`  /cc @michaelb990

File: docs/proposals/PROPOSAL_E.md [hunk 5]
```diff
@@ -80,17 +79,21 @@ Extend the Image Manifest with a refers field (existing registries should ignore
 {
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.manifest.v1+json",
```
Comment (mounchin) [inline docs/proposals/PROPOSAL_E.md:81]: Should this be `application/vnd.oci.artifact.manifest.v1+json` ? Or is there a scenario where an image manifest can refer another image manifest?

**Comment by jdolitsky** [comment]
I've gone ahead and resolved merge conflicts / committed suggested changes to keep this moving. @michaelb990 - is this ready for review/merge?

**Review by jdolitsky — APPROVED** [review]

**Comment by sudo-bmitch** [comment]
> I've gone ahead and resolved merge conflicts / committed suggested changes to keep this moving. @michaelb990 - is this ready for review/merge?

On the WG meeting last week, I think we ended up unsure if this made sense to implement since it would result in added complexity for registry servers that would reduce the ability to cache the output, and clients already have client-side filtering implemented to support the tag output and any registries that don't support filtering. Lets save this for a merge or close decision on tomorrow's meeting.

**Comment by michaelb990** [comment]
@sudo-bmitch 
>On the WG meeting last week, I think we ended up unsure if this made sense to implement since it would result in added complexity for registry servers that would reduce the ability to cache the output, and clients already have client-side filtering implemented to support the tag output and any registries that don't support filtering. Lets save this for a merge or close decision on tomorrow's meeting.

* Does the fact that it is a SHOULD not a MUST help with the complexity worries?
* The recent updates to the annotations should solve the caching issues.
* I think not having server-side filtering is one of the big drawbacks of the tagging scheme, so I'm hoping we see things move to server-side solutions "soon". 

Will post to a vote.

**Review by sajayantony — APPROVED** [review]
LGTM for merging once the vote is in. Any minor adjustments can be made post merge in my opinion.

**Comment by sudo-bmitch** [comment]
> * Does the fact that it is a SHOULD not a MUST help with the complexity worries?

I'd lean towards MAY, but that's not a strong opinion.

> * The recent updates to the annotations should solve the caching issues.

Thank you!

> * I think not having server-side filtering is one of the big drawbacks of the tagging scheme, so I'm hoping we see things move to server-side solutions "soon". 
> 
> Will post to a vote.

The big advantage of having the API to me isn't filtering, but the lack of client side maintenance of the tag, and elimination of race conditions. (Plus it declutters the tag listing.) Thanks for getting the vote out. 👍

**Comment by dlorenc** [comment]
+1 here too. Thanks!

**Comment by rchincha** [comment]
lgtm

**Review by sudo-bmitch — APPROVED** [review]

**Pull request merged**